### ISSUE TITLE
Config: Some resets don't set the default value

### ISSUE THREAD

**Issue by rayfordshire** [neomutt/neomutt#3741]
# The Problem

Some config variables have only a runtime default which is not known at compile time, examples include `$tmp_dir` (default is $TMPDIR with fallback to `/tmp`), `$suspend` (default is `yes` or `no` depending on whether NeoMutt is process group leader), `$editor` (default $EDITOR + fallbacks) and probably some more.

Here is an example which shows that default is not equal to resetted:
`# TMPDIR=foo neomutt -F /dev/null`
`:set tmp_dir?` -> `foo`
`reset tmp_dir`
`:set tmp_dir?` -> `/tmp`

# Idea 1: "store tuples"

Extend the Config system to store an additional default value which can be set like a normal value, e.g. `set_default("tmp_dir", "my/fancy/dir");` with fall back to the compile time default.  When NeoMutt initialises itself, it sets the defaults according to whatever it found out.

# Idea 2: "default functions"

Extend Config system by replacing the default value with a function (pointer) returning the default value.  Whenever the default is requested that function is asked and it returns whatever it currently finds is the default value. This approach allows to dynamically react to changes in the environment and adapt the config value accordingly, e.g. the default value should be the current time.
The current defaults would just be a function which returns set default, e.g. `default_auto_tag() { return false; }`

# Idea 3: "layers"

(sort of othogonal to 1 and 2)
Extend the Config system so that a ConfigSet falls back to another ConfigSet if the key is not found in it.  Think of them as layers:
top layer are all configs which the user set, bottom layer are all the "defaults".  Setting a default variable would be setting the value in the lower layer.  This would also allow to detect if a user actively set a variable or not by the fact whether it is present in the upper level or not. (More layers are easily possible.)

---

@flatcap what do you think of these approaches?  Which one would fit in your vision?

Idea 1 and 3 are more in line with "config system is only a data storage".  Idea 2 seems more flexible.

**Comment by flatcap** [comment]
> Some resets don't set the default value

*sigh*
Fixed in #3742 

All the magic happens here in `mutt_init()`:
https://github.com/neomutt/neomutt/blob/8c3731301728c9f8b276ab50d8d3a1da7653bb3f/init.c#L603-L678

The complete list is: `$charset`, `$editor`, `$from`, `$hostname`, `$mailcap_path`, `$real_name`, `$spool_file`, `$suspend`, `$tmp_dir`.

**Comment by flatcap** [comment]
> Idea 1: "store tuples"

This isn't necessary.
The config system already has the facility to override the values coded into `struct ConfigDef MainVars[]`, etc by calling `cs_str_initial_set()` and `cs_str_reset()`.

> Idea 2: "default functions"
> a function (pointer) returning the default value.

This is already possible with `cs_str_initial_get()` / `cs_he_initial_get()`.

> This approach allows to dynamically react to changes in the environment

There aren't any changes to the environment.
When a program is started it gets a copy of its parent's environment.
That environment can only be changed by NeoMutt.

Yes, there's a `setenv` command in NeoMutt, but that only changes the environment that's given to children, e.g. mailcap programs.
After `mutt_init()` has initialised the config, NeoMutt only refers to its config variables.

> Idea 3: "layers"

That's already implemented and tested in the config code -- the "layers" are the `ConfigSubset`s.
See: [Config System](https://neomutt.org/dev/config) -- (updated today)

As yet, they're unused, but the plan is for `Account` and `Mailbox` objects to have their own `ConfigSubset` to allow Account- and Mailbox-specific config, without the need for `account-hook`s and `folder-hook`s.

The config will look like:
```sh
account Apple 
    set folder     = "imaps://apple.org/"
    set imap_user  = "apple"
    set imap_pass  = "AAA"
    set spool_file = +INBOX
    set postponed  = "+drafts"
    set trash      = "+trash"'                                                                                                  
    
account Banana    
    set folder     = "imap://banana.org/"   
    set imap_user  = "banana"  
    set imap_pass  = "BBB"               
    set spool_file = +INBOX  
    set postponed  = "+drafts"  
    set trash      = "+trash"' 
```

The indentation is cosmetic.
An `account` command scopes all the following config until another `account` command or the end-of-file.
